src/libraries/System.Memory.Data/src/System/BinaryData.cs

```diff
@@ -0,0 +1,262 @@
+        /// </summary>
+        /// <param name="data">The array to wrap.</param>
+        /// <returns>A wrapper over <paramref name="data"/>.</returns>
+        public static BinaryData FromBytes(byte[] data) => new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from a string by converting
+        /// the string to bytes using the UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <returns>A value representing the UTF-8 encoding of <paramref name="data"/>.</returns>
+        public static BinaryData FromString(string data) => new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static BinaryData FromStream(Stream stream)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, false).GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static Task<BinaryData> FromStreamAsync(Stream stream, CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, true, cancellationToken);
+        }
+
+        private static async Task<BinaryData> FromStreamAsync(Stream stream, bool async, CancellationToken cancellationToken = default)
+        {
+            int streamLength = 0;
+            if (stream.CanSeek)
+            {
+                long longLength = stream.Length - stream.Position;
+                if (longLength > int.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(stream), SR.ArgumentOutOfRange_StreamLengthMustBeInt32);
+                }
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
+            if (async)
+            {
+                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
```

Comment:
We need to pass in CopyToBufferSize explicitly because this lib is targeting .NET Standard 2.0 which doesn't have the overload that doesn't take a buffer size?

Reply:
I believe that to be the case: https://apisof.net/catalog/System.IO.Stream.CopyToAsync(Stream,CancellationToken)

Reply:
Correct, this overload is not present.
https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=net-5.0#System_IO_Stream_CopyToAsync_System_IO_Stream_System_Threading_CancellationToken_

Do you think this library should copy https://github.com/dotnet/runtime/blob/5c65d891f203618245184fa54397ced0a8ca806c/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L114?  We could factor that into a helper method.

Reply:
Essentially that's what the logic above that starts the method is trying to do, except that it would then also be used as the argument to CopyTo{Async} in addition to being used to configure the MemoryStream.

Reply:
True, so.
```C#
int bufferSize = canSeek ? Math.Min(streamLength, CopyToBufferSize) : CopyToBufferSize;
```

And in this case I'll provide to CopyTo as well, since it's better than the default on desktop.